### PR TITLE
red-dev-avalanche-chainlink-tutorial (Round-2)

### PR DESCRIPTION
**Pull request by tamil-reddev**
This pull request is for red·dev's entry "avalanche-chainlink-tutorial" for Round 2 of the Avalanche Tutorials contest. This entry was submitted originally under the name of red·dev's CEO, Christopher David Sturgeon ("Kit"). It is a group effort by part of the red·dev team. Today, team member Tamil Selvan is submitting the request. We look forward to your critique and comments! Thank you.

### CODE REVIEW

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/README.md [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+
+* jsonparse, ethuint256 & ethtx are core adaptors and are executed synchronously. For more details, you will find documentation for each adapter's usage [here](https://docs.chain.link/docs/core-adapters/).
+* runlog - By adding the address parameter, you make the event filter of the RunLog initiator more restrictive, only listening for events from that address, instead of any address. By adding the requesters parameter, you only allow requests to come from an address within the array. Please have a look at this [video](https://www.youtube.com/watch?v=ZB3GLtQvgME&38m38s) to learn how to get your oracle address.
+
+
+# Smart contracts
+
+This section will explain the smart contracts we have written for communicating with the Chainlink node through oracle address and job id.
+
+Below is the code snippet which you can copy and paste into the Remix IDE and then compile and deploy.
```
Comment (jpop32) [inline build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/README.md:334]: Could you please expand this section? 'Compile and deploy' are pretty complex steps with a lot of possibilities for mistakes (solidity versions, compile options)...

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/README.md [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+
+```
```
Comment (jpop32) [inline build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/README.md:336]: Could you please add format definition to all code sections? Like:
\`\`\`solidity
soliditiy code goes here
\`\`\`
It helps with syntax highlighting

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/README.md [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+    assembly { // solhint-disable-line no-inline-assembly
+      result := mload(add(source, 32))
+    }
+  }
+
+}
+```
+Deployed smart contracts:
+
+![smartcontracts-deployed](img/smartContract-deploy.png)
```
Comment (jpop32) [inline build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/README.md:394]: Please rename all the image files so they have a common and unique prefix that can identify them as belonging to this tutorial (GitBook tool we use puts all the images in a single direcory, this helps them be grouped together). For example
chainlink-tutorial-00-name-of-first-image.png,
chainlink-tutorial-01-name-of-second-image.png,
...

**Comment by jpop32** [comment]
Really excellent work, congrats! Could you please add a summary / conclusion to the tutorial?

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/README.md (hunk 1 above)
Comment (tamil-reddev) [reply build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/README.md]: Hi, @jpop32 I have added additional details at 4ba5be59729a9bc2d8bf6a840c768ba00355cd4b

**Comment by tamil-reddev** [comment]
> Really excellent work, congrats! Could you please add a summary / conclusion to the tutorial?

Thanks. Added conclusion section please have a look at 4ba5be59729a9bc2d8bf6a840c768ba00355cd4b

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/README.md (hunk 3 above)
Comment (tamil-reddev) [reply build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/README.md]: Updated image names as suggested. Please have a look at 4ba5be59729a9bc2d8bf6a840c768ba00355cd4b

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/README.md (hunk 2 above)
Comment (tamil-reddev) [reply build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/README.md]: Added as suggested. Please have a look at 4ba5be59729a9bc2d8bf6a840c768ba00355cd4b

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/ansible-chainlink-avalanche-setup/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# reddev infrastructure automation repository
```
Comment (cgcardona) [inline build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/ansible-chainlink-avalanche-setup/README.md:1]: [Headings should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.23.1/doc/Rules.md#md022)

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/ansible-chainlink-avalanche-setup/README.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+Setup requirements:
+1. Ansible v2.9.9 or later
+2. Get Vultr.com API key from https://my.vultr.com/settings/#settingsapi
+3. Create file at ~/.vultr.ini and fill in your API key like so:
+```
```
Comment (cgcardona) [inline build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/ansible-chainlink-avalanche-setup/README.md:6]: What data type is this block of code?

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/chainlink-external-adaptor/app.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import conf from 'nconf';
```
Comment (cgcardona) [inline build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/chainlink-external-adaptor/app.js:1]: Port js -> ts

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/chainlink-external-adaptor/json-rpc-client.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+		        } else if (jsonRPCRequest.id !== undefined) {
+				        return Promise.reject(new Error(response.statusText));
+				    }
+}));
+
+const api = () => {
+	    return {
+		            post: (method, params, chain) => {
+				                console.log(`Request for chain ${chain}`);
+				                switch (chain) {
```
Comment (cgcardona) [inline build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/chainlink-external-adaptor/json-rpc-client.js:27]: Use 2 spaces for indentation

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/chainlink-external-adaptor/package.json [hunk 8]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (cgcardona) [inline build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/chainlink-external-adaptor/package.json:1]: Remove package.json and package-lock.json

**Review by cgcardona — CHANGES_REQUESTED** [review]
Minor changes requested

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/ansible-chainlink-avalanche-setup/README.md (hunk 4 above)
Comment (tamil-reddev) [reply build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/ansible-chainlink-avalanche-setup/README.md:1]: This is done. @cgcardona

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/chainlink-external-adaptor/json-rpc-client.js (hunk 7 above)
Comment (tamil-reddev) [reply build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/chainlink-external-adaptor/json-rpc-client.js:27]: This is done @cgcardona

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/chainlink-external-adaptor/package.json (hunk 8 above)
Comment (tamil-reddev) [reply build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/chainlink-external-adaptor/package.json]: Removed the files. But, can you explain why don't we need the files?

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/chainlink-external-adaptor/app.js (hunk 6 above)
Comment (tamil-reddev) [reply build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/chainlink-external-adaptor/app.js:1]: do u want me to change the entire project to ts? or only the app.js? @cgcardona

File: build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/ansible-chainlink-avalanche-setup/README.md (hunk 5 above)
Comment (tamil-reddev) [reply build/tutorials/tutorials-contest/red-dev-avalanche-chainlink-tutorial/ansible-chainlink-avalanche-setup/README.md:7]: This is not kind of any coding. This is our VPS environment where we installed the ansible v2.9.9 and executed the automated scripts.

**Comment by kitpub** [comment]
As I just wrote in our other tutorial's pull request (and will gladly repeat again!), we are so honored to have won and to have had this chance to contribute to Avalanche! Thank you so much for the opportunity. The team at red·dev is looking forward to much more to come as the ecosystem grows, and we are 100% committed to its success. Please let us know how we can help going forward.

**Comment by kitpub** [comment]
One follow-up question regarding this tutorial in particular. We learned how to build a Chainlink adapter because we saw it as the only way to query the P chain from the C chain--something we need to do in our RediYeti dapp since it monitors validators' performance. Of course, since the two chains are right there on each node running side-by-side, we thought, wouldn't it be nice if the C-chain could natively query the P-chain, skipping Chainlink? It would seem to me that perhaps a precompiled contract in coreth might be the way to go. Do you know if the coreth team would be open to this feature or to making a change another way that would accomplish the same thing?

**Pull request merged**